### ISSUE TITLE
Pipeline Caching - .net - Why `packages.lock,json` file is required?

### ISSUE THREAD

**Issue by tymtam2** [MicrosoftDocs/azure-devops-docs#6806]
This is about This is about https://docs.microsoft.com/en-us/azure/devops/pipelines/caching/index?view=azure-devops#netnuget

I don't believe that the information provided is enough to implement caching for .NET. for a Pipeline.

The documentation is: 

_If you use PackageReferences to manage NuGet dependencies directly within your project file and have a packages.lock.json file, you can enable caching by setting the NUGET_PACKAGES environment variable to a path under $(Pipeline.Workspace) and caching this directory._

Here are the first hurdle I'm dealing with:
**Do you really need packages.lock.json for caching to work? If so, why?** 

I used the following key nuget | "$(Agent.OS)" | packages

My assumption is that the way it works is: 
1. Check for cache and get data if data exists
2. (As a separate tasks) nuget restore:
  a) if nothing has changed in the projects and there are no new packages out there then no changes
  b) if projects changed or there are new packages restore updates the packages folder (this is the same as on my dev machine)
3. Build runs 
4. The cache value is updated in the Post-job

This means there is no need for a lock file. 

If this thinking is correct then having a dynamic key based on the hash of the lock file(s) would result in multiple caches when the lockfile changes and that would be wasteful since the old values will never be used. This makes me think I am not understanding this correctly. 

In summary, I think the documentation could be extended to explain more how the caching works, what quotas there are, what to do in case of multiple lock files, what if there is no lockfile, etc.

**Comment by msebolt** [comment]
@tymtam2 Thank you for the feedback, assigning this to the author for review.

**Comment by kmadof** [comment]
Any news here?

**Comment by cxwtool** [comment]
This issue hasn't been updated in more than 180 days, so we've closed it. If you feel the issue is still relevant and needs fixed, please reopen it and we'll take another look. We appreciate your feedback and apologize for any inconvenience.

**Comment by grantbowering** [comment]
This was passed off to @chasewilson and then @ramiMSFT, but was never addressed or acknowledged, and timed out with no resolution.  The documentation still provides no guidance for how to achieve pipeline caching of nuget dependencies when there's more than one package.lock.json in the universe (which I'd guess is about 90% of cases.  How often does one have a single .NET project in a CI pipeline? 😕)

**Comment by ramiMSFT** [comment]
Apologies for the delayed response. The content of the packages.lock.json will be hashed to produce a dynamic cache key. This will guarantee the uniqueness of the cache because every time a new dependency is added, your packages.lock.json will be different. See [Configure the Cache task](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#configure-the-cache-task) for details. 
If you want to use multiple packages.lock.json, you can still use the following example without making any changes. The content of all the packages.lock.json files will be hashed and if one of the files are changed a new cache key will be generated.

```yaml
variables:
  NUGET_PACKAGES: $(Pipeline.Workspace)/.nuget/packages

steps:
- task: Cache@2
  inputs:
    key: 'nuget | "$(Agent.OS)" | $(Build.SourcesDirectory)/**/packages.lock.json'
    restoreKeys: |
       nuget | "$(Agent.OS)"
       nuget
    path: $(NUGET_PACKAGES)
  displayName: Cache NuGet packages
```
